database/user.go

```diff
@@ -224,26 +229,75 @@ func (db *DB) UserByStripeID(ctx context.Context, id string) (*User, error) {
 	return &u, nil
 }
 
+// UserBySub returns the user with the given sub. If `create` is `true` it will
+// create the user if it doesn't exist. The sub is the Kratos id of that user.
+func (db *DB) UserBySub(ctx context.Context, sub string, create bool) (*User, error) {
+	users, err := db.managedUsersByField(ctx, "sub", sub)
+	if create && errors.Contains(err, ErrUserNotFound) {
+		_, email, err := jwt.UserDetailsFromJWT(ctx)
+		if err != nil {
+			// Log the error but don't do anything differently.
+			db.staticLogger.Debugf("We failed to extract the expected user infotmation from the JWT token. Error: %s", err.Error())
+		}
+		u, err := db.UserCreate(ctx, email, "", sub, TierFree)
+		// If we're successful or hit any error, other than a duplicate key we
+		// want to just return. Hitting a duplicate key error means we ran into
+		// a race condition and we can easily recover from that.
+		if err == nil || !strings.Contains(err.Error(), "E11000 duplicate key error collection") {
```

Comment:
Are keys for new users not randomly generated? When would we run into a duplicate key error?

Reply:
This error is handled in two functions - `UserBySub` and `UserByEmail`.

For `UserBySub` this duplicate key error is possible because the user might have registered with Kratos and then might use the same JWT to make simultaneous request to two different servers. Those can cause a data race resulting in this duplicate key error. We recover from that by just fetching the user from the DB.

For `UserByEmail` the situation is a little different. The `email` field isn't set to be unique right now. I just noticed that because of your comment. We should make that unique because we use it as username. I cannot make it unique right now because we have plenty of users with empty emails in the DB. First we need to do the full data migration and then make emails unique. And after we make them unique we will be in the very same situation where we can get a dup and we can just fetch from the DB.

F/U: #70

Reply:
F/U: #71